modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassSystemMenu.java

```diff
@@ -148,7 +160,33 @@ private void insertMenu(final Menu parent, final MenuBase mb, int pos) {
 
         final FilteredList<MenuItemBase> filteredItems = items.filtered(x -> x.isVisible());
 
-        filteredItems.addListener((ListChangeListener.Change<? extends MenuItemBase> change) -> {
+        ListChangeListener menuItemListener = createListener(glassMenu);
+        filteredItems.addListener(menuItemListener);
+        menuListeners.put(glassMenu, menuItemListener);
+        listenerItems.put(menuItemListener, filteredItems);
+
+        for (MenuItemBase item : items) {
+            if (item instanceof MenuBase) {
+                // submenu
+                addMenu(glassMenu, (MenuBase)item);
+            } else {
+                // menu item
+                addMenuItem(glassMenu, item);
+            }
+        }
+        glassMenu.setPixels(getPixels(mb));
+
+        setMenuBindings(glassMenu, mb);
+
+        if (parent != null) {
+            parent.insert(glassMenu, pos);
+        } else {
+            glassSystemMenuBar.insert(glassMenu, pos);
+        }
+    }
+
+    private ListChangeListener createListener(final Menu glassMenu) {
+        ListChangeListener<MenuItemBase> answer = ((ListChangeListener.Change<? extends MenuItemBase> change) -> {
```

Comment:
minor: I see extra parenthesis around the whole, and I think you can just `return` this immediately (no need for `answer` local)

Reply:
I typically add a local variable for this (which is removed by javac) to make it easier to add debugging, but I'll remove it.